syncing/src/test/kotlin/maru/syncing/beaconchain/CLSyncServiceImplTest.kt

```diff
@@ -0,0 +1,493 @@
+
+    createBlocks(
+      beaconChain = beaconChain2,
+      genesisBeaconBlock = genesisBeaconBlock,
+      genesisTimestamp = genesisBeaconBlock.beaconBlock.beaconBlockHeader.timestamp,
+      validators = validators,
+      signatureAlgorithm = signatureAlgorithm,
+      keypair = keypair,
+    )
+    peerLookup = spy(p2pNetwork1.getPeerLookup())
+    clSyncService =
+      CLSyncServiceImpl(
+        beaconChain = beaconChain1,
+        executorService = Executors.newCachedThreadPool(),
+        validatorProvider = StaticValidatorProvider(validators),
+        allowEmptyBlocks = true,
+        peerLookup = peerLookup,
+        besuMetrics = TestMetricsSystemAdapter,
+        metricsFacade = TestMetricsFacade,
+        pipelineConfig = Config(blocksBatchSize = 10u, blocksParallelism = 1u),
+      )
+
+    try {
+      p2pNetwork1.start()
+      p2pNetwork2.start()
+      p2pNetwork1.addStaticPeer(createPeerAddress(port2))
+
+      awaitUntilAsserted { assertNetworkHasPeers(network = p2pNetwork1, peers = 1) }
+      awaitUntilAsserted { assertNetworkHasPeers(network = p2pNetwork2, peers = 1) }
+    } catch (e: Exception) {
+      p2pNetwork1.stop()
+      p2pNetwork2.stop()
+      throw IllegalStateException("Failed to start P2P networks", e)
+    }
+  }
+
+  @AfterEach
+  fun tearDown() {
+    p2pNetwork1.stop()
+    p2pNetwork2.stop()
+  }
+
+  @Test
+  fun `chain sync downloads and imports blocks from another node`() {
+    val synced = AtomicBoolean(false)
+    clSyncService.setSyncTarget(100uL)
+    clSyncService.onSyncComplete { synced.set(true) }
+    awaitUntilAsserted { assertThat(synced).isTrue() }
+    assertThat(beaconChain1.getLatestBeaconState().latestBeaconBlockHeader.number).isEqualTo(100uL)
+    assertThat(beaconChain1.getLatestBeaconState()).isEqualTo(beaconChain2.getLatestBeaconState())
+    for (i in 1uL..100uL) {
+      assertThat(beaconChain1.getSealedBeaconBlock(i)).isEqualTo(beaconChain2.getSealedBeaconBlock(i))
+      assertThat(beaconChain1.getBeaconState(i)).isEqualTo(beaconChain2.getBeaconState(i))
+    }
+  }
+
+  @Test
+  fun `sync target can be updated ahead of current target and continue downloading`() {
+    // sync to block 50
+    val synced = AtomicBoolean(false)
+    clSyncService.setSyncTarget(50uL)
+    clSyncService.onSyncComplete { synced.set(true) }
+    awaitUntilAsserted { assertThat(synced).isTrue() }
+
+    assertThat(beaconChain1.getLatestBeaconState().latestBeaconBlockHeader.number).isEqualTo(50uL)
+    assertThat(beaconChain1.getLatestBeaconState()).isEqualTo(beaconChain2.getBeaconState(50uL))
+    for (i in 1uL..50uL) {
+      assertThat(beaconChain1.getSealedBeaconBlock(i)).isEqualTo(beaconChain2.getSealedBeaconBlock(i))
+      assertThat(beaconChain1.getBeaconState(i)).isEqualTo(beaconChain2.getBeaconState(i))
+    }
+
+    // update sync target to 100
+    synced.set(false)
+    clSyncService.setSyncTarget(100uL)
+    awaitUntilAsserted { assertThat(synced).isTrue() }
```

Comment:
I feel like these 3 are always together and could be factored out and reused. And maybe even the `val synced = AtomicBoolean(false)` can be defined inside this new function